### ISSUE TITLE
GenericstTest failing to compile

### ISSUE THREAD

**Issue by rubencgt** [dropwizard/dropwizard#2089]
Can not compile the application due to this Error.
the class is GenericsTest

`Error:(59, 28) java: method getTypeParameter in class io.dropwizard.util.Generics cannot be applied to given types;
  required: java.lang.Class<?>,java.lang.Class<? super T>
  found: java.lang.Class<capture#1 of ?>,java.lang.Class<capture#2 of ?>
  reason: cannot infer type-variable(s) T
    (argument mismatch; java.lang.Class<capture#2 of ?> cannot be converted to java.lang.Class<? super T>)`

**Comment by nickbabcock** [comment]
What version of java are you using to compile?

Can you try updating to a newer compiler?

**Comment by isaki-x** [comment]
I have seen cases from time to time where javac compiler will allow a particular generic declaration/implied cast but IDEs such as Eclipse and IntelliJ do not. This is usually fixable with minor code changes so that things work everywhere.

**Comment by nickbabcock** [comment]
CC PR author: @manuel-hegner to see if they have any insight.

**Comment by manuel-hegner** [comment]
Is this error still present after you switch IntelliJ to the  eclipse
compiler?

Nick Babcock <notifications@github.com> schrieb am Do., 29. Juni 2017,
18:08:

**Comment by nickbabcock** [comment]
Switching to the eclipse compiler fixed the issue.

Is there anyway we can get those tests to compile with out-of-the-box intellij?

**Comment by manuel-hegner** [comment]
I'll look into it tomorrow, but I have no IntelliJ.

Nick Babcock <notifications@github.com> schrieb am Do., 29. Juni 2017,
19:39:

**Comment by FredDeschenes** [comment]
I've found a way to get this to compile in both Eclipse and IntelliJ, currently running the tests to see if everything is still good (although it shouldn't change anything at runtime). I'll send a PR if everything still works.

**Comment by manuel-hegner** [comment]
I have no idea why this hels with the problem. Sometimes IntelliJ is weird this way I guess.

But nice if this works :grinning:

**Comment by FredDeschenes** [comment]
I'd say Java generics are weird ;)

**Comment by isaki-x** [comment]
I've never seen a parametrized unit test before. I also have no idea why that fixes the issue.

**Comment by FredDeschenes** [comment]
The parametized test just tells JUnit to create multiple instances of the test, using the values provided by the `@Parameters` as constructor arguments for the instance. For `GenericsTest`, this means that 6 instances of the test are created, being fed by the data returned by the `GenericsTest::data` function.

The issue itself has nothing to do with this though, it's just "ye olde Java generics" acting differently depending on which compiler is used. In this case, `Generics::getTypeParameter(Class<?> klass, Class<? super T> bound)` expects `bound` to be a `Class<? super T>`, but in this case, the IntelliJ compiler was seeing `bound` as a `Class<?>`, which at runtime would make no difference (as Java generics aren't reified and the `T` itself isn't really relevant for `getTypeParameter`), but in theory could cause some troubles, so the compiler wouldn't let it go through.

We used to get weird issues like this at work a couple of years ago when some people used Eclipse and others used IntelliJ. Usually, Eclipse's compiler is the one that "whines" the most about those issues, this one just happened to annoy IntelliJ's compiler instead.

TLDR : As I mentioned above, Java generics are weird, and the tiny differences in how the different compilers handle them are annoying (considering that most of the time, it makes absolutely no difference at runtime).

**Comment by arteam** [comment]
Thank you for the explanation!. Closed via #2095.
